### PR TITLE
[SYCL] Emit llvm.compiler.used for internal device globals

### PR DESCRIPTION
**Pull request by Fznamznon**
This prevents early optimizations from removing device globals with
internal linkage.

### CODE REVIEW

**Review by premanandrao — DISMISSED** [review]
Changes look okay to me.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 1]
```diff
@@ -2157,6 +2157,16 @@ void CodeGenModule::SetCommonAttributes(GlobalDecl GD, llvm::GlobalValue *GV) {
         VD->getStorageDuration() == SD_Static)
       addUsedOrCompilerUsedGlobal(GV);
   }
+
+  if (getLangOpts().SYCLIsDevice) {
```
Comment (smanna12) [inline clang/lib/CodeGen/CodeGenModule.cpp:2161]: Changes look good to me. Could you please add a comment for future reference? Thanks

Comment (Fznamznon) [reply clang/lib/CodeGen/CodeGenModule.cpp:2161]: Sure, done.

**Review by steffenlarsen — APPROVED** [review]
LGTM! 🎉

**Review by smanna12 — APPROVED** [review]
Thank you.

**Review by premanandrao — APPROVED** [review]

**Comment by Fznamznon** [comment]
The following failures in OCL x64 LLVM Test Suite (pull_request_target) test run look like a common problem and therefore unrelated to this PR:
```
********************
Failed Tests (1):
  SYCL :: XPTI/kernel/content.cpp

********************
Unexpectedly Passed Tests (2):

  SYCL :: SubGroup/reduce_spirv13.cpp
  SYCL :: SubGroup/reduce_spirv13_fp64.cpp
2 warning(s) in tests
```

**Review by smanna12 — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]

**Comment by Fznamznon** [comment]
@intel/llvm-gatekeepers , I believe this one is ready for merge.

**Pull request merged**